mail_notification_custom_subject/models/mail_message_custom_subject.py

```diff
@@ -0,0 +1,243 @@
+    )
+    mako_template_env.globals.update(
+        {
+            "str": str,
+            "quote": urls.url_quote,
+            "urlencode": urls.url_encode,
+            "datetime": datetime,
+            "len": len,
+            "abs": abs,
+            "min": min,
+            "max": max,
+            "sum": sum,
+            "filter": filter,
+            "reduce": functools.reduce,
+            "map": map,
+            "round": round,
+            # dateutil.relativedelta is an old-style class and cannot be directly
+            # instanciated wihtin a jinja2 expression, so a lambda "proxy" is
+            # is needed, apparently.
+            "relativedelta": lambda *a, **kw: relativedelta.relativedelta(*a, **kw),
+        }
+    )
+    mako_safe_template_env = copy.copy(mako_template_env)
+    mako_safe_template_env.autoescape = False
+except ImportError:
+    _logger.warning("jinja2 not available, templating features will not work!")
+
+
+class MailMessageCustomSubject(models.Model):
+    _name = "mail.message.custom.subject"
+    _description = "Mail Message Custom Subject"
+
+    name = fields.Char(string="Template Name")
+    model_id = fields.Many2one(
+        comodel_name="ir.model",
+        string="Models",
+        required=True,
+        help="Model(s) where this template applies",
+    )
+    subtype_ids = fields.Many2many(
+        comodel_name="mail.message.subtype",
+        string="Subtype",
+        required=True,
+        help="Message subtypes followed,"
+        " meaning subtypes that will be pushed onto the user's Wall",
+    )
+    subject_template = fields.Char(
+        string="Subject Template",
+        required=True,
+        help="Subject (placeholders may be used here)",
+    )
+    position = fields.Selection(
+        selection=[
+            ("append_before", _("Append Before")),
+            ("append_after", _("Append After")),
+            ("replace", _("Replace")),
+        ],
+        string="Position",
+        default="replace",
+        help="Whether to replace, append at beggining or append at end to other"
+        " templates that apply to a given context",
+    )
+
+    @api.multi
+    def render_subject_template(self, model, res_id):
+        subject_override = ""
+        for template in self:
+            rendered_subject_templates = template._render_template(
+                template_txt=template.subject_template, model=model, res_ids=res_id
+            )
+            if template.position == "replace":
+                subject_override = rendered_subject_templates
+            elif template.position == "append_before":
+                subject_override = rendered_subject_templates + subject_override
+            elif template.position == "append_after":
+                subject_override += rendered_subject_templates
+        return subject_override or False
+
+    @api.multi
+    def _render_template(self, template_txt, model, res_ids, post_process=False):
```

Comment:
Aren't you able to call existing methods in mail.template or similar to avoid to rewrite all this stuff?